### ISSUE TITLE
Check database migration algorithm

### ISSUE THREAD

**Issue by mmitech** [ethcore/parity#1848]
I woke up to my node (mining pool) having a local fork, I've restarted parity to solve the problem but parity always crashes on block verification :
```
2016-08-05 10:36:08 Stage 3 block verification failed for #2004423 (36a7…8f66)
Error: Block(UnknownParent(54606fe3b0a45c111d713dccca5b59fffff0615420f7aa113d3470a74ac436c3))
```
Edit: sorry forget to tell the version: v1.3.0-unstable-21c65a9-20160801 on Ubuntu server 14.04
```
2016-08-05 10:35:47 Starting Parity/v1.3.0-unstable-21c65a9-20160801/x86_64-linux-gnu/rustc1.10.0
2016-08-05 10:35:47 Using state DB journalling strategy fast
2016-08-05 10:35:49 Configured for Frontier/Homestead using Ethash engine
2016-08-05 10:35:52 Updated conversion rate to Ξ1 = US$10.88 (21883752000 wei/gas)
2016-08-05 10:35:57 Public node URL: enode://34b7171026274aa7e5e88d2ee1edbb514e768a0d49a92e0f761e129aa63f9c88001ce01c24eae6349866a62d28f03cae2c4db06eeddfe8ded65c6c3742f8fc43@2                              12.93.226.167:30303
2016-08-05 10:36:06 Syncing #2004420 c087…56b2      0 blk/s    0 tx/s   0 Mgas/s       0+    0 Qed   #2004420    1/ 2/25 peers     50 MiB db   88 KiB chain    2 KiB queue    8                               KiB sync
thread 'IO Worker #2' panicked at 'Not found! 487f97c89801cc230dd59fba34c1397fa2e322024412aab28a9ad1cc6a39875e', .cargo/git/checkouts/parity-6ba620efeffb2216/master/util/src/t                              rie/triedb.rs:215
note: Run with `RUST_BACKTRACE=1` for a backtrace.
2016-08-05 10:36:08 Finishing work, please wait...
2016-08-05 10:36:08 Stage 3 block verification failed for #2004423 (36a7…8f66)
Error: Block(UnknownParent(54606fe3b0a45c111d713dccca5b59fffff0615420f7aa113d3470a74ac436c3))
```

**Comment by gavofyork** [comment]
could you post the full logs?

**Comment by mmitech** [comment]
done

**Comment by gavofyork** [comment]
yes - looks like it has a corrupted database now. do you still have the logs from the point of it mining on its own fork? when you restarted it, how did you shut it down?

**Comment by gavofyork** [comment]
also, for production environments, especially when using the beta/unstable branches, i recommend syncing with `--pruning=archive`.

**Comment by mmitech** [comment]
Sorry! it is too late for that since I've replaced 1.3 with the stable 1.2.3 (which worked fine), so I can't reproduce the issue, I have logs however:
 
```
[1;30m2016-08-03 17:14:14 [0m   [1;36m 0[0m/[1;36m48[0m/[1;36m50[0m peers   [1;34m  50 MiB[0m db [1;34m  87 MiB[0m chain [1;34m   2 KiB[0m queue [1;34m  21 KiB[0m sync
[1;30m2016-08-03 17:14:34 [0mImported [1;37m#2004412[0m [1;37m8cdf…fa12[0m ([1;33m11[0m txs, [1;33m0.30[0m Mgas, [1;35m43.82[0m ms, [1;34m2.04[0m KiB)
[1;30m2016-08-03 17:14:45 [0m   [1;36m 0[0m/[1;36m49[0m/[1;36m50[0m peers   [1;34m  50 MiB[0m db [1;34m  87 MiB[0m chain [1;34m   2 KiB[0m queue [1;34m  21 KiB[0m sync
[1;30m2016-08-03 17:14:51 [0mImported [1;37m#2004413[0m [1;37ma5ca…731e[0m ([1;33m5[0m txs, [1;33m0.11[0m Mgas, [1;35m29.65[0m ms, [1;34m1.14[0m KiB)
[1;30m2016-08-03 17:15:16 [0m   [1;36m 0[0m/[1;36m49[0m/[1;36m50[0m peers   [1;34m  50 MiB[0m db [1;34m  87 MiB[0m chain [1;34m   2 KiB[0m queue [1;34m  21 KiB[0m sync
[1;30m2016-08-03 17:15:19 [0mImported [1;37m#2004414[0m [1;37mf1c6…bcfa[0m ([1;33m14[0m txs, [1;33m0.30[0m Mgas, [1;35m34.40[0m ms, [1;34m2.07[0m KiB)
[1;30m2016-08-03 17:15:36 [0mImported [1;37m#2004415[0m [1;37mddf0…9b5e[0m ([1;33m1[0m txs, [1;33m0.02[0m Mgas, [1;35m4.31[0m ms, [1;34m0.63[0m KiB)
[1;30m2016-08-03 17:15:47 [0m   [1;36m 0[0m/[1;36m49[0m/[1;36m50[0m peers   [1;34m  50 MiB[0m db [1;34m  87 MiB[0m chain [1;34m   2 KiB[0m queue [1;34m  21 KiB[0m sync
[1;30m2016-08-03 17:15:48 [0mImported [1;37m#2004416[0m [1;37m731b…c545[0m ([1;33m42[0m txs, [1;33m0.91[0m Mgas, [1;35m107.63[0m ms, [1;34m5.21[0m KiB)
[1;30m2016-08-03 17:16:06 [0mImported [1;37m#2004417[0m [1;37m061e…6f5f[0m ([1;33m13[0m txs, [1;33m0.31[0m Mgas, [1;35m44.38[0m ms, [1;34m2.02[0m KiB)
[1;30m2016-08-03 17:16:17 [0m   [1;36m 0[0m/[1;36m49[0m/[1;36m50[0m peers   [1;34m  50 MiB[0m db [1;34m  87 MiB[0m chain [1;34m   2 KiB[0m queue [1;34m  21 KiB[0m sync
[1;30m2016-08-03 17:16:21 [0mImported [1;37m#2004418[0m [1;37md18d…2701[0m ([1;33m21[0m txs, [1;33m0.47[0m Mgas, [1;35m153.58[0m ms, [1;34m2.84[0m KiB)
[1;30m2016-08-03 17:16:48 [0m   [1;36m 0[0m/[1;36m49[0m/[1;36m50[0m peers   [1;34m  50 MiB[0m db [1;34m  87 MiB[0m chain [1;34m   2 KiB[0m queue [1;34m  21 KiB[0m sync
[1;30m2016-08-03 17:17:08 [0mImported [1;37m#2004419[0m [1;37m4549…a444[0m ([1;33m18[0m txs, [1;33m0.53[0m Mgas, [1;35m53.19[0m ms, [1;34m2.70[0m KiB)
[1;30m2016-08-03 17:17:19 [0m   [1;36m 0[0m/[1;36m49[0m/[1;36m50[0m peers   [1;34m  50 MiB[0m db [1;34m  88 MiB[0m chain [1;34m   2 KiB[0m queue [1;34m  21 KiB[0m sync
[1;30m2016-08-03 17:17:20 [0mImported [1;37m#2004420[0m [1;37mc087…56b2[0m ([1;33m36[0m txs, [1;33m1.00[0m Mgas, [1;35m85.53[0m ms, [1;34m4.96[0m KiB)
thread 'IO Worker #2' panicked at 'Not found! 487f97c89801cc230dd59fba34c1397fa2e322024412aab28a9ad1cc6a39875e', .cargo/git/checkouts/parity-6ba620efeffb2216/master/util/src/trie/triedb.rs:215
note: Run with `RUST_BACKTRACE=1` for a backtrace.
[1;30m2016-08-03 17:17:28 [0mFinishing work, please wait...

[1;30m2016-08-03 17:17:33 [0mStarting [1;37mParity/v1.3.0-unstable-21c65a9-20160801/x86_64-linux-gnu/rustc1.10.0[0m
[1;30m2016-08-03 17:17:33 [0mUsing state DB journalling strategy [1;37mfast[0m
[1;30m2016-08-03 17:17:35 [0mConfigured for [1;37mFrontier/Homestead[0m using [1;33mEthash[0m engine
[1;30m2016-08-03 17:17:43 [0mUpdated conversion rate to Ξ1 = [1;37mUS$9.44[0m ([1;33m25221954000[0m wei/gas)
[1;30m2016-08-03 17:17:43 [0mPublic node URL: [1;37menode://34b7171026274aa7e5e88d2ee1edbb514e768a0d49a92e0f761e129aa63f9c88001ce01c24eae6349866a62d28f03cae2c4db06eeddfe8ded65c6c3742f8fc43@212.93.226.167:30303[0m
thread 'IO Worker #0' panicked at 'Not found! 487f97c89801cc230dd59fba34c1397fa2e322024412aab28a9ad1cc6a39875e', .cargo/git/checkouts/parity-6ba620efeffb2216/master/util/src/trie/triedb.rs:215
note: Run with `RUST_BACKTRACE=1` for a backtrace.
[1;30m2016-08-03 17:17:46 [0mFinishing work, please wait...
[1;30m2016-08-03 17:17:46 [0mStage 3 block verification failed for #2004423 (36a7…8f66)
Error: Block(UnknownParent(54606fe3b0a45c111d713dccca5b59fffff0615420f7aa113d3470a74ac436c3))

[1;30m2016-08-03 17:17:48 [0mStarting [1;37mParity/v1.3.0-unstable-21c65a9-20160801/x86_64-linux-gnu/rustc1.10.0[0m
[1;30m2016-08-03 17:17:48 [0mUsing state DB journalling strategy [1;37mfast[0m
[1;30m2016-08-03 17:17:49 [0mConfigured for [1;37mFrontier/Homestead[0m using [1;33mEthash[0m engine
[1;30m2016-08-03 17:17:53 [0mUpdated conversion rate to Ξ1 = [1;37mUS$9.44[0m ([1;33m25221954000[0m wei/gas)
[1;30m2016-08-03 17:17:57 [0mPublic node URL: [1;37menode://34b7171026274aa7e5e88d2ee1edbb514e768a0d49a92e0f761e129aa63f9c88001ce01c24eae6349866a62d28f03cae2c4db06eeddfe8ded65c6c3742f8fc43@212.93.226.167:30303[0m
thread 'IO Worker #3' panicked at 'Not found! 487f97c89801cc230dd59fba34c1397fa2e322024412aab28a9ad1cc6a39875e', .cargo/git/checkouts/parity-6ba620efeffb2216/master/util/src/trie/triedb.rs:215
note: Run with `RUST_BACKTRACE=1` for a backtrace.
[1;30m2016-08-03 17:17:59 [0mFinishing work, please wait...
[1;30m2016-08-03 17:17:59 [0mStage 3 block verification failed for #2004423 (36a7…8f66)
Error: Block(UnknownParent(54606fe3b0a45c111d713dccca5b59fffff0615420f7aa113d3470a74ac436c3))
[1;30m2016-08-03 17:17:59 [0mBad header 2004424 (1b22…e200) from 129: Geth/v1.4.10-stable/windows/go1.6.2, state = ChainHead

[1;30m2016-08-03 17:18:01 [0mStarting [1;37mParity/v1.3.0-unstable-21c65a9-20160801/x86_64-linux-gnu/rustc1.10.0[0m
[1;30m2016-08-03 17:18:01 [0mUsing state DB journalling strategy [1;37mfast[0m
[1;30m2016-08-03 17:18:02 [0mConfigured for [1;37mFrontier/Homestead[0m using [1;33mEthash[0m engine
[1;30m2016-08-03 17:18:10 [0mPublic node URL: [1;37menode://34b7171026274aa7e5e88d2ee1edbb514e768a0d49a92e0f761e129aa63f9c88001ce01c24eae6349866a62d28f03cae2c4db06eeddfe8ded65c6c3742f8fc43@212.93.226.167:30303[0m
thread 'IO Worker #2' panicked at 'Not found! 487f97c89801cc230dd59fba34c1397fa2e322024412aab28a9ad1cc6a39875e', .cargo/git/checkouts/parity-6ba620efeffb2216/master/util/src/trie/triedb.rs:215
note: Run with `RUST_BACKTRACE=1` for a backtrace.
[1;30m2016-08-03 17:18:11 [0mFinishing work, please wait...
[1;30m2016-08-03 17:18:11 [0mStage 3 block verification failed for #2004423 (36a7…8f66)
Error: Block(UnknownParent(54606fe3b0a45c111d713dccca5b59fffff0615420f7aa113d3470a74ac436c3))

```

**Comment by gavofyork** [comment]
there are two possible reasons this was caused; either a failed database upgrade to 1.3 or a pruned chain caused by a bug fixed in https://github.com/ethcore/parity/pull/1816 .

if you had been running 1.3 for a reasonable period of time without a problem then it was likely the latter bug. if it happened soon after upgrading (i.e. within the first 1000 or so blocks) then it could have been a bad database migration.

**Comment by mmitech** [comment]
not really sure since I've upgraded a couple of days ago, but I did a fresh reindex to test the speed of fast syncing.

**Comment by gavofyork** [comment]
what steps did you follow to do a "fresh reindex"?

**Comment by mmitech** [comment]
renamed the folder .parity and started my parity daemon.

**Comment by gavofyork** [comment]
so to be clear, this is the order of what happened:

1. running 1.2.3 fine
2. a couple of days ago: upgraded to 1.3.0 (running fine).
3. one day ago: renamed .parity and rerun 1.3.0
4. parity failed with above error

did you check how long it took to sync parity? do you have the logs from the sync process?

**Comment by mmitech** [comment]
This was the order:
1- running 1.2.3 fine
2- a couple of days ago: renamed .parity, compiled and upgraded to 1.3.0 (running fine).
3- parity failed with above error

Syncing took a bit longer than I expected (a couple of hours), the first million block was really fast then it started slowing down, I will try to send you logs. (I could resync and give you results if you are collecting data).


BTW, I forgot to mention this before and I am not sure if this is what caused the issue or if it is related at all, but I run parity as an upstart service, when it locally forked I restarted it as I've mentioned before, but it was throwing the "Too many open files error" which didn't beofre

```
^[[1;30m2016-08-05 10:23:52 ^[[0mStarting ^[[1;37mParity/v1.3.0-unstable-21c65a9-20160801/x86_64-linux-gnu/rustc1.10.0^[[0m
^[[1;30m2016-08-05 10:23:52 ^[[0mUsing state DB journalling strategy ^[[1;37mfast^[[0m
thread '<main>' panicked at 'called `Result::unwrap()` on an `Err` value: Io(Error { repr: Os { code: 24, message: "Too many open files" } })', ../src/libcore/result.rs:785
note: Run with `RUST_BACKTRACE=1` for a backtrace.
```
I didn't know that upstart doesn't look at /etc/security/limits.conf so I added the limits to my upstart config and then got the error in this issue after that.
